### PR TITLE
[JENKINS-59046] - Human readable permissions for JCasC

### PR DESCRIPTION
**Pull request by casz**
- Added export tests
- Allows human readable of permission same style as role and matrix auth

### CODE REVIEW

File: src/test/resources/io/jenkins/plugins/folderauth/casc/config.yml [hunk 1]
```diff
@@ -27,32 +27,32 @@ jenkins:
...
-            - id: "hudson.model.Item.Workspace"
-            - id: "hudson.model.Computer.Provision"
-            - id: "hudson.model.Item.WipeOut"
-            - id: "hudson.model.View.Read"
-            - id: "hudson.model.View.Create"
-            - id: "hudson.model.Item.Delete"
-            - id: "hudson.model.Computer.ExtendedRead"
-            - id: "hudson.model.Computer.Configure"
-            - id: "hudson.model.Computer.Disconnect"
+            - id: "View/Delete"
```
Comment (casz) [inline src/test/resources/io/jenkins/plugins/folderauth/casc/config.yml]: Only converted some of the permission to show we can handle both.

File: src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Implements lookup for {@link Permission}s.
+ */
+// Imported from https://github.com/jenkinsci/configuration-as-code-plugin/blob/727c976d137461f146b301f302d1552ca81de75e/plugin/src/main/java/io/jenkins/plugins/casc/util/PermissionFinder.java
```
Comment (oleg-nenashev) [inline src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java]: Could we make it a public API there?

Comment (casz) [reply src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java]: It gone and removed 😆

Comment (casz) [reply src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java]: Also Matrix auth and Role auth both have them locally

Comment (casz) [reply src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java]: We could add it back make it public and hope that both role, folder and matrix will accept the pull request 😆

**Comment by casz** [comment]
hmm tests worked earlier but have failed to cleanup after wondering for a bit

**Comment by AbhyudayaSharma** [comment]
```
Expected: is "agentRoles:\n- agents:\n - \"agent1\"\n name: \"agentRole1\"\n permissions:\n - id: \"Agent/Configure\"\n - id: \"Agent/Disconnect\"\n sids:\n - \"user1\"\nfolderRoles:\n- folders:\n - \"root\"\n name: \"viewRoot\"\n permissions:\n - id: \"Job/Read\"\n sids:\n - \"user1\"\nglobalRoles:\n- name: \"admin\"\n permissions:\n - id: \"View/Delete\"\n - id: \"Agent/Connect\"\n - id: \"Agent/Create\"\n - id: \"View/Configure\"\n - id: \"Job/Configure\"\n - id: \"Agent/Build\"\n - id: \"Job/Cancel\"\n - id: \"Job/Read\"\n - id: \"Agent/Delete\"\n - id: \"Job/Build\"\n - id: \"Job/ExtendedRead\"\n - id: \"SCM/Tag\"\n - id: \"Job/Move\"\n - id: \"Job/Discover\"\n - id: \"Overall/Read\"\n - id: \"Job/Create\"\n - id: \"Job/Workspace\"\n - id: \"Agent/Provision\"\n - id: \"Job/WipeOut\"\n - id: \"View/Read\"\n - id: \"View/Create\"\n - id: \"Job/Delete\"\n - id: \"Agent/ExtendedRead\"\n - id: \"Agent/Configure\"\n - id: \"Overall/Administer\"\n - id: \"Agent/Disconnect\"\n sids:\n - \"admin\"\n- name: \"read\"\n permissions:\n - id: \"Overall/Read\"\n sids:\n - \"user1\"\n"
but: was "agentRoles:\n- agents:\n - \"agent1\"\n name: \"agentRole1\"\n permissions:\n - id: \"Agent/Configure\"\n - id: \"Agent/Disconnect\"\n sids:\n - \"user1\"\nfolderRoles:\n- folders:\n - \"root\"\n name: \"viewRoot\"\n permissions:\n - id: \"Job/Read\"\n sids:\n - \"user1\"\nglobalRoles:\n- name: \"read\"\n permissions:\n - id: \"Overall/Read\"\n sids:\n - \"user1\"\n- name: \"admin\"\n permissions:\n - id: \"Job/Move\"\n - id: \"Job/Build\"\n - id: \"View/Create\"\n - id: \"Agent/Configure\"\n - id: \"Job/Read\"\n - id: \"Agent/Create\"\n - id: \"Job/ExtendedRead\"\n - id: \"Job/Delete\"\n - id: \"Agent/Build\"\n - id: \"View/Configure\"\n - id: \"Agent/Provision\"\n - id: \"SCM/Tag\"\n - id: \"Job/Create\"\n - id: \"Job/Discover\"\n - id: \"Agent/ExtendedRead\"\n - id: \"Agent/Connect\"\n - id: \"Agent/Delete\"\n - id: \"Agent/Disconnect\"\n - id: \"Job/Cancel\"\n - id: \"Overall/Read\"\n - id: \"Job/WipeOut\"\n - id: \"Overall/Administer\"\n - id: \"View/Delete\"\n - id: \"Job/Configure\"\n - id: \"Job/Workspace\"\n - id: \"View/Read\"\n sids:\n - \"admin\"\n"
```

It looks like the exported roles are sorted in a different order. The text starts differing from index 300.

**Comment by casz** [comment]
Running into [JENKINS-17200](https://issues.jenkins-ci.org/browse/JENKINS-17200)

File: src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java (hunk 2 above)
Comment (oleg-nenashev) [reply src/main/java/io/jenkins/plugins/folderauth/misc/PermissionFinder.java]: Yeah, I think 3 copy-pastes definitely justify having an API method in JCasC or in the core

**Comment by oleg-nenashev** [comment]
I wonder how to make Exporter to support a fancy format by default when we generalize this API

**Comment by AbhyudayaSharma** [comment]
Thanks a lot @casz ! This is awesome 🎉

**Pull request merged**